### PR TITLE
Restrict cosine_similarity max value to 1.0

### PR DESCRIPTION
**Pull request by v0dro**
This clamps the highest value achievable by `cosine_similarity` to `1.0`.

Fix for https://github.com/pytorch/pytorch/issues/29442

CC: @ezyang

### CODE REVIEW

**Comment by v0dro** [comment]
A test for this already seems to exist here: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L6323

Should I write an additional test to check for the data posted in #29442 ?

**Comment by vadimkantorov** [comment]
If I understand correctly, this lead to zero gradient if cosine similarity happens to be numerically above 1.0, right?

**Comment by ezyang** [comment]
Needs test

**Comment by albanD** [comment]
@ezyang to avoid the gradient issue mentioned above, we can do do:
```cpp
w12.div_(n12);
{
  // This is only to avoid spurious values > 1.0
  // We do not want the gradient for these values to be 0.
  NoGradGuard guard;
  w12.clamp_max_(1.0);
}
return w12;
```

Do you think it is a sensible thing to do?

**Comment by ezyang** [comment]
@albanD seems plausible.

btw @v0dro, the test you cited wasn't failing prior to your patch. So it doesn't seem like it's sufficient to test the particular bug here. And I'd probably also want to test the gradient behavior too.

**Comment by v0dro** [comment]
@ezyang the test specified in the bug report uses some particular data read from rather large files, and looking at the test cases I don't think its advisable to use this data directly inside a test. In this case, how can I best import the data that is needed for the test into the pytorch codebase? I'm not sure if there's a way to reproduce the bug using a smaller dataset.

**Comment by albanD** [comment]
Hi,
From what I remember in that example, they provide a whole matrix while only one entry is problematic.
Then you can check how many elements from that entry are actually needed to see the error? Do you think it will be too many of them?

**Comment by ezyang** [comment]
Yes, what Alban said.

**Comment by nikitaved** [comment]
Hey, maybe it would also make sense to clamp_min at -1.0?

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 8942db1:
* **3/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8942db16f613794a0816aa98f7738ffc28ec2b13).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3956569) pytorch_xla_linux_xenial_py3_6_clang7_build (1/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74509947))
<details>
<summary>
<code>Dec 15 05:05:28 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 05:05:28 +++ eval 'extract_trap_cmd ' 
Dec 15 05:05:28 ++++ extract_trap_cmd 
Dec 15 05:05:28 ++++ printf '%s\n' '' 
Dec 15 05:05:28 +++ printf '%s\n' cleanup 
Dec 15 05:05:28 ++ trap -- ' 
Dec 15 05:05:28 cleanup' EXIT 
Dec 15 05:05:28 ++ which sccache 
Dec 15 05:05:28 ++ sccache --stop-server 
Dec 15 05:05:28 Stopping sccache server... 
Dec 15 05:05:28 error: couldn't connect to server 
Dec 15 05:05:28 caused by: Connection refused (os error 111) 
Dec 15 05:05:28 ++ true 
Dec 15 05:05:28 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 05:05:28 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Dec 15 05:05:28 ++ true 
Dec 15 05:05:28 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 05:05:28 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 05:05:28 ++ RUST_LOG=sccache::server=error 
Dec 15 05:05:28 ++ sccache --start-server 
Dec 15 05:05:28 Starting sccache server... 
Dec 15 05:05:28 ++ sccache --zero-stats 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3956721) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (2/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74518070))
<details>
<summary>
<code>Dec 15 06:27:45 ../.local/lib/python3.6/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%]</code>
</summary>

```
Dec 15 06:27:39 ============================= test session starts ============================== 
Dec 15 06:27:39 platform linux -- Python 3.6.9, pytest-5.2.1, py-1.8.0, pluggy-0.13.0 -- /usr/bin/python3.6 
Dec 15 06:27:39 cachedir: .pytest_cache 
Dec 15 06:27:39 rootdir: /var/lib/jenkins 
Dec 15 06:27:39 plugins: hypothesis-3.59.0, sugar-0.9.2 
Dec 15 06:27:39 collecting ... WARNING:root:This caffe2 python run does not have GPU support. Will run in CPU only mode. 
Dec 15 06:27:39 INFO:caffe2.python.net_drawer:Cannot import pydot, which is required for drawing a network. This can usually be installed in python with "pip install pydot". Also, pydot requires graphviz to convert dot files to pdf: in ubuntu, this can usually be installed with "sudo apt-get install graphviz". 
Dec 15 06:27:45 net_drawer will not run correctly. Please install the correct dependencies. 
Dec 15 06:27:45 collected 2894 items / 1 skipped / 2893 selected 
Dec 15 06:27:45  
Dec 15 06:27:45 ../.local/lib/python3.6/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%] 
Dec 15 06:27:45  
Dec 15 06:27:45 =================================== FAILURES =================================== 
Dec 15 06:27:45 ________________________ TestAllCompare.test_allcompare ________________________ 
Dec 15 06:27:45  
Dec 15 06:27:45 self = <caffe2.python.allcompare_test.TestAllCompare testMethod=test_allcompare> 
Dec 15 06:27:45  
Dec 15 06:27:45     @given( 
Dec 15 06:27:45 >       d=st.integers(1, 5), n=st.integers(2, 11), num_procs=st.integers(1, 8) 
Dec 15 06:27:45     ) 
Dec 15 06:27:45     def test_allcompare(self, d, n, num_procs): 
```

</details>

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| CircleCI <sup>pytorch_linux_xenial_py3_6_gcc5_4_build</sup> | <sup>Build</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by nikitaved** [comment]
@v0dro , I do believe that if similarity measure exceeds 1.0, it might get lower than -1.0. Also, technically, hard thresholding does modify (sub)gradients. Therefore, I believe it makes sense to test how the patch affects the gradient behaviour when the input vectors live in small and high dimensional spaces.

**Comment by albanD** [comment]
@nikitaved as mentioned above, I don't think we actually want the gradients to be impacted by this change. So clamping should be done in a no grad block.

**Comment by nikitaved** [comment]
@albanD I totally agree, but the forward is different now. I do not mean to modify backward, but I believe it makes sense to test whether the function does have somewhat closer gradients with/without the patch. Why? Suppose, for simplicity, that `x, y` live in **huge** dimensional space and are already normalized. The cosine similarity is just an inner product then: 
`f(x, y) = <x, y>`. Also, `df/dy = x`. Suppose we minimize some scalar loss `L` and we backprop to `y`. Then `y` receives `dL/df * x` gradient when we do not clamp, and `dL/df_clamped * x` when we clamp. Suppose further that `y = Az`. Then, after backproping to `z` we would have `A^T * dL/df * x` vs. `A^T * dL/df_clamped * x`, and these gradients might be quite different, especially when in high dimensions. 

Maybe there are some safeguards I am not aware of. But I am actually quite curious how PyTorch handles numerical stuff like this. Is there anywhere I could read about it?

**Comment by albanD** [comment]
I'm not sure to understand your point. My suggestion was to return the exact same gradient as before.
Given that we only modify the forward up to floating point precision. I don't think there is a need to change the backward formula.

In particular, because this function is autodiffed, we want to make sure the clamp happens in a no grad block to make sure that we don't force these gradients to 0 but we keep them at their current value.

**Comment by nikitaved** [comment]
My point is that, if you keep the formulas for the gradients the same, but change forwards even slightly, it might cause a dramatic change in the gradient values.

**Comment by nikitaved** [comment]
Also, I did check the implementation of the `cosine_similarity`. It is not numerically stable...
It first finds the inner product `<x,y>`, and only then does the normalization. A much better approach is to first normalize inputs, and only then compute the inner product. What do you think about this? Then maybe then the fix is not necessary.

**Comment by albanD** [comment]
I don't think any particular care has been taken to try and make this function more numerically stable.
Maybe simply re-ordering the operations would be sufficient indeed.

**Comment by nikitaved** [comment]
Ok, then I will play with this and submit PR if it works out just fine...

**Comment by nikitaved** [comment]
@albanD , could you please help me out? Do you think this code is correct? I am particularly interested in that `unsqueeze` part. Thank you!
```python
Tensor x1_norm = at::sum(x1 * x1, dim).clamp_min_(eps).sqrt_();
Tensor x1_normalized = x1.div(x1_norm.unsqueeze(dim));
```

**Comment by albanD** [comment]
I think you want to use the `at::norm()` function directly. It has a better behavior at `0` (https://github.com/pytorch/pytorch/pull/2775).
The `unsqueeze()` is fine here. A better way is to pass `keepdim=True` to the reduction function you use so that you don't need to `unsqueeze()` afterwards.

**Comment by nikitaved** [comment]
@albanD, thank you! Still struggling with the last test... Anyway, would just having `c.max() < 1.0` resolve the issue [29442](https://github.com/pytorch/pytorch/issues/29442)? If so, then just improving the method's numerical stability does the trick. I just to do something about the last test, or ignore it. It works on the border of numerical precision...

**Comment by albanD** [comment]
If the new implementation ensures that we never return something > 1 then that will work.
If it is still "unlikely" I think we want to keep the clamp to ensure this is always true.

**Comment by rgommers** [comment]
Let's put this PR on hold, I think gh-31378 is the preferred fix.

**Comment by rgommers** [comment]
> If the new implementation ensures that we never return something > 1 then that will work.
> If it is still "unlikely" I think we want to keep the clamp to ensure this is always true.

This was answered in the other PR by the way - it's "never" by design.
